vitrina/smart_contracts/views.py

```diff
@@ -4,30 +4,169 @@
+        super().setup(request, *args, **kwargs)
+        self.object = get_object_or_404(
+            Project.public.all().prefetch_related(
+                Prefetch(
+                    "datasets",
+                    queryset=Dataset.public.all().order_by("organization_id"),
+                    to_attr="public_datasets",
+                )
+            ),
+            pk=kwargs["pk"],
+        )
+
+
+class AgreementListView(
+    LoginRequiredMixin,
+    PermissionRequiredMixin,
+    BaseAgreementMixin,
+    HistoryMixin,
+    TemplateView,
+):
+    model = Agreement
+    template_name = "smart_contracts/agreement_list.html"
+
+    detail_url_name = "project-detail"
+    history_url_name = "project-history"
+
+    object: Project
+
+    def has_permission(self) -> bool:
+        return has_perm(self.request.user, Action.VIEW, self.object)
+
+    def get_context_data(self, **kwargs: Any) -> dict:
+        context = super().get_context_data(**kwargs)
+
+        project_agreements = (
+            Agreement.objects.filter(
+                project=self.object,
+            )
+            .order_by("-created_at")
+            .all()
+        )
+
+        paginator = Paginator(project_agreements, 10)
+        page_number = self.request.GET.get("page")
+        page = paginator.get_page(page_number)
+
+        context.update(
+            {
+                "project": self.object,
+                "agreements": page.object_list,
+                "agreement_status_descriptions": AGREEMENT_STATUS_DESCRIPTIONS,
+                "page_obj": page,
+                "paginator": paginator,
+                "can_update_project": has_perm(
+                    self.request.user, Action.UPDATE, self.object
+                ),
+                "can_view_agreements": has_perm(
+                    self.request.user, Action.VIEW, Agreement, self.object
+                ),
+                "parent_links": {
+                    reverse("home"): _("Pradžia"),
+                    reverse("project-list"): _("Panaudojimo atvejai"),
+                    reverse("project-detail", args=[self.object.pk]): self.object,
+                    None: _("Sutartys"),
+                },
+            }
+        )
+
+        return context
+
+
+class AgreementDetailView(
+    LoginRequiredMixin,
+    PermissionRequiredMixin,
+    BaseAgreementMixin,
+    HistoryMixin,
+    TemplateView,
+):
+    model = Agreement
+    template_name = "smart_contracts/agreement_detail.html"
+
+    detail_url_name = "project-detail"
+    history_url_name = "project-history"
+
+    object: Project
+    agreement: Agreement
+
+    def has_permission(self) -> bool:
+        return has_perm(self.request.user, Action.VIEW, self.agreement)
+
+    def setup(self, request: WSGIRequest, *args: Any, **kwargs: Any) -> None:
+        super().setup(request, *args, **kwargs)
+        self.agreement = get_object_or_404(
+            Agreement.objects.all()
```

Comment:
`.all()` usually redundant if followed by another methods.

Reply:
I'd keep it

> explicit is better than implicit